index.d.ts

```diff
@@ -52,15 +43,18 @@ export interface AnyAction extends Action {
  *
  * *Do not put API calls into reducers.*
  *
- * @template S State object type.
+ * @template S The type of state consumed and produced by this reducer.
+ * @template A The type of actions the reducer can potentially respond to.
  */
-export type Reducer<S> = (state: S, action: AnyAction) => S;
+export type Reducer<S = {}, A extends Action = Action> = (state: S, action: A) => S;
```

Comment:
To me making types that don't agree with the runtime is an abuse of type system.

However, I can see that this approach gives really nice DX, so I tend to agree with you. It should be up to developer to decide the level of strictness they want.